src/commands/scp.ts

```diff
@@ -0,0 +1,151 @@
+    "scp <source> <destination>",
+    // TODO: support scp across multiple remote hosts.
+    "SCP copies files between a local and remote host.",
+    (yargs) =>
+      yargs
+        .positional("source", {
+          type: "string",
+          demandOption: true,
+          description: "Format: [hostname:]file or file",
+        })
+        .positional("destination", {
+          type: "string",
+          demandOption: true,
+          description: "Format: [hostname:]file or file",
+        })
+        .option("i", {
+          alias: "identity",
+          type: "string",
+          describe:
+            "Selects a file from which the identity (private key) for public key authentication is read",
+          demandOption: true,
+        })
+        .option("r", {
+          alias: "recursive",
+          type: "boolean",
+          describe: "Recursively copy entire directories",
+        })
+        .option("reason", {
+          describe: "Reason access is needed",
+          type: "string",
+        })
+        .option("account", {
+          type: "string",
+          describe: "The account on which the instance is located",
+        })
+        .option("sudo", {
+          type: "boolean",
+          describe: "Add user to sudoers file",
+        }),
+    guard(scpAction)
+  );
+
+/** Transfers files between a local and remote hosts using SSH.
+ *
+ * Implicitly gains access to the SSH resource if required.
+ */
+const scpAction = async (args: yargs.ArgumentsCamelCase<ScpCommandArgs>) => {
+  const authn = await authenticate();
+
+  const host = getHostIdentifier(args.source, args.destination);
+
+  if (!host) {
+    throw "Could not determine host identifier from source or destination";
+  }
+
+  const requestId = await provisionRequest(authn, args, host);
+
+  if (!requestId) {
+    throw "Server did not return a request id. Please contact support@p0.dev for assistance.";
+  }
+
+  if (!args.identity) {
+    throw "Please provide a path to the identity file";
+  }
+
+  const identity = await fs.readFile(args.identity, "utf8");
+  if (!identity) {
+    throw "Could not read identity file, please check the path and try again.";
+  }
+
+  const result = await fetchExerciseGrant(authn, {
+    requestId,
+    destination: host,
+    // Generates the public key from the private key which prevents having to pass the public key separately.
+    // Works with rsa/ed25519 algorithms and keys generated with ssh-keygen or openssl.
+    publicKey: sshpk.parseKey(identity, "pem").toString("ssh"),
+  });
+
+  // replace the host with the linuxUserName@instanceId
+  const { source, destination } = replaceHostWithInstance(result, args);
+
+  await scp(authn, result, {
+    ...args,
+    source,
+    destination,
+  });
+};
+
+/** If a path is not explicitly local, use this pattern to determine if it's remote */
+const REMOTE_PATTERN_COLON = /^([^:]+:)(.*)$/; // Matches host:[path]
+
+const isExplicitlyRemote = (path: string): boolean => {
```

Comment:
One thing to note is that detecting the remote / local paths is not complete. SCP has more complex logic around that. E.g. there could be colons in the local path which you can specify by starting path with `.`, `..`, or `/`

Reply:
Good point, I created a ticket for expanding on the more complex logic for SCP local and remote paths.

https://linear.app/p0-security/issue/ENG-1931/improve-remote-host-and-local-host-checking-for-scp-requests